src/functor.jl

```diff
@@ -296,8 +340,8 @@ fmapstructure(f, x; kwargs...) = fmap(f, x; walk = (f, x) -> map(f, children(x))
     fcollect(x; exclude = v -> false)
 
 Traverse `x` by recursing each child of `x` as defined by [`functor`](@ref)
-and collecting the results into a flat array, ordered by a breadth-first
-traversal of `x`, respecting the iteration order of `children` calls.
+and collecting the results into a flat array, ordered by a depth-first,
+post-order traversal of `x` that respects the iteration order of `children` calls.
```

Comment:
This is technically breaking (as can be seen from the test changes), but it's unclear to me how many users were relying on the breadth-first iteration order. The one example I found on JuliaHub, AlphaZero, was not. It's also odd that the original worked breadth-first when pretty much all other traversals based on functors are depth-first. So RFC, and happy to back this out if we consider it too breaking.

Reply:
I think it's fine to have this breaking change.

Reply:
I think that nothing now in Optimisers.jl now cares about the order taken by `fmap`. 

Of course the two halves of `destructure` care a lot and must match. I haven't completely understood how #31 would have to change; it currently uses `fmap` for one half and something hand-written for the other.

Reply:
The order in `fmap` has always been stable, things brings `fcollect` in line as it was the odd one out.